### PR TITLE
571 model instance unit test

### PR DESCRIPTION
**Pull request by jsadler2**
@pkdash please revise.

### CODE REVIEW

**Comment by hydrobuild** [comment]
Can one of the admins verify this patch?

File: hs_modelinstance/tests/test_modelinstance_metadata.py [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+
+    def tearDown(self):
+        super(TestModelInstanceMetaData, self).tearDown()
+        if os.path.exists(self.temp_dir):
+            shutil.rmtree(self.temp_dir)
+
+    def test_allowed_file_types(self):
+        # test allowed file type is '.*'
+        self.assertIn('.*', ModelInstanceResource.get_supported_upload_file_types())
+        self.assertEquals(len(ModelInstanceResource.get_supported_upload_file_types()), 2)
```
Comment (pkdash) [inline hs_modelinstance/tests/test_modelinstance_metadata.py]: Does this test pass? Number of supported file types should be 1.

File: hs_modelinstance/tests/test_modelinstance_metadata.py [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+
+        # there should not be any content file
+        self.assertEquals(self.resModelInstance.files.all().count(), 0)
+
+        # Upload any file type should pass both the file pre add check post add check
+        files = [UploadedFile(file=self.text_file_obj, name=self.text_file_obj.name)]
+        utils.resource_file_add_pre_process(resource=self.resModelInstance, files=files, user=self.user,
+                                            extract_metadata=False)
```
Comment (pkdash) [inline hs_modelinstance/tests/test_modelinstance_metadata.py]: Should also test for ```extract_metadata=True```

File: hs_modelinstance/tests/test_modelinstance_metadata.py [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+
+        utils.resource_file_add_process(resource=self.resModelInstance, files=files, user=self.user,
+                                        extract_metadata=False)
+
+        # there should one content file
+        self.assertEquals(self.resModelInstance.files.all().count(), 1)
+
+    def test_extended_metadata_CRUD(self):
+        # test the core metadata at this point
```
Comment (pkdash) [inline hs_modelinstance/tests/test_modelinstance_metadata.py]: What about core metadata elements like 'rights', 'relation', and 'creator'?

File: hs_modelinstance/tests/test_modelinstance_metadata.py [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+
+        # there shouldn't any contributor element
+        self.assertEquals(self.resModelInstance.metadata.contributors.all().count(), 0)
+
+        # check that there are no extended metadata elements at this point
+        self.assertEquals(self.resModelInstance.metadata.model_output, None)
+        self.assertEquals(self.resModelInstance.metadata.executed_by, None)
+
+        # create
+        self.resModelInstance.metadata.create_element('ModelOutput')
```
Comment (pkdash) [inline hs_modelinstance/tests/test_modelinstance_metadata.py]: Also need to test passing the parameter ```includes_output=True``` and ```includes_output=False``` for creating ```ModelOutput``` metadata element.

File: hs_modelinstance/tests/test_modelinstance_metadata.py [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+
+        modeloutput_element = self.resModelInstance.metadata.model_output
+        self.assertEquals(modeloutput_element.includes_output, False)
+
+        # multiple ModelOutput elements are not allowed - should raise exception
+        with self.assertRaises(IntegrityError):
+            self.resModelInstance.metadata.create_element('ModelOutput', includes_output=True)
```
Comment (pkdash) [inline hs_modelinstance/tests/test_modelinstance_metadata.py]: Also test with ```includes_output=False```

File: hs_modelinstance/tests/test_modelinstance_metadata.py [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+        executedby_element = self.resModelInstance.metadata.executed_by
+        self.assertEquals(executedby_element.model_name, self.resModelProgram.metadata.title.value)
+        self.assertEquals(executedby_element.model_program_fk, self.resModelProgram)
+
+        # multiple ExecutedBy elements are not allowed - should raise exception
+        with self.assertRaises(IntegrityError):
+            self.resModelInstance.metadata.create_element('ExecutedBy', model_name=self.resModelProgram2.short_id)
+
+        # update
+        self.resModelInstance.metadata.update_element('ModelOutput', self.resModelInstance.metadata.model_output.id,
```
Comment (pkdash) [inline hs_modelinstance/tests/test_modelinstance_metadata.py]: Test update also with ```includes_output=False```

File: hs_modelinstance/tests/test_modelinstance_metadata.py [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+    def test_get_xml(self):
+        self.resModelInstance.metadata.create_element('Description', abstract="test abstract")
+        self.resModelInstance.metadata.create_element('Subject', value="test subject")
+        self.resModelInstance.metadata.create_element('ModelOutput', includes_output=True)
+        self.resModelInstance.metadata.create_element('ExecutedBy', model_name=self.resModelProgram.short_id)
+
+        # test if xml from get_xml() is well formed
+        ET.fromstring(self.resModelInstance.metadata.get_xml())
+
+    def test_metadata_on_content_file_delete(self):
```
Comment (pkdash) [inline hs_modelinstance/tests/test_modelinstance_metadata.py]: Need to test for core metadata elements like ```contributors```, ```rights```, and ```relation```

**Comment by jsadler2** [comment]
@pkdash we have revised our tests according to your review. please verify our changes

**Comment by mohamedmorsy** [comment]
@pkdash Done.

**Comment by pkdash** [comment]
+1

**Pull request merged**